PSReadLine/Suggestion.cs

```diff
@@ -0,0 +1,61 @@
+using System;
+
+namespace Microsoft.PowerShell
+{
+    public partial class PSConsoleReadLine
+    {
+        private class SuggestionModeRestore : IDisposable
+        {
+            private bool oldSuggestionMode;
+
+            internal SuggestionModeRestore(bool showSuggestion)
+            {
+                oldSuggestionMode = _singleton._showSuggestion;
+                _singleton._showSuggestion = showSuggestion;
+            }
+
+            public void Dispose()
+            {
+                _singleton._showSuggestion = oldSuggestionMode;
+            }
+        }
+
+        private string _suggestionText;
+        private uint _lastRenderedTextHash;
+        private bool _showSuggestion = true;
+
+        /// <summary>
+        /// Turn off the prediction feature for the scope of the calling method.
+        /// Note: the calling method need to use this method with the 'using' statement/variable,
+        /// so that the suggestion feature can be properly restored.
+        /// </summary>
+        private SuggestionModeRestore PredictionOff() => new SuggestionModeRestore(false);
+
+        private string GetSuggestion(string text)
+        {
+            if (_options.PredictionStyle == PredictionStyle.None || !_showSuggestion ||
+                LineIsMultiLine() || string.IsNullOrWhiteSpace(text))
+            {
+                _suggestionText = null;
+                _lastRenderedTextHash = 0;
+                return null;
+            }
+
+            try
+            {
+                uint textHash = _lastRenderedTextHash;
+                _lastRenderedTextHash = FNV1a32Hash.ComputeHash(text);
+                if (textHash != _lastRenderedTextHash)
```

Comment:
Why not just compare strings? Hashing has to look at every character but string comparison can return quickly, e.g. if lengths don't match or at the first mismatch.

Reply:
`_lastRenderedTextHash` is not really that useful. It saves a call to `GetHistorySuggestion` only in some cases where text is pasted to the console. I removed it for now.